[5.6] Add ability to override default api endpoint

Miguel-Serejo opened the pull request:
Fixes #23864 by allowing a user to set a "mail.endpoints.sparkpost" config key to choose between available endpoints found at https://developers.sparkpost.com/api/

example addition to config/mail.php:
```
/*
|--------------------------------------------------------------------------
| Mail Service Endpoints
|--------------------------------------------------------------------------
|
| When using an email service, you might need to override the default API
| endpoints. This allows you to configure different endpoints based on your
| environment.
|
*/
'endpoints' => [
  'sparkpost' =>  env('SPARKPOST_ENDPOINT', 'https://api.sparkpost.com/api/v1'),
],
```

I'm not sure how to test this, but I would argue that any test to this feature would just be a test of the config helper. 

I'm also unsure if other mail drivers should get the same treatment as I don't use any of them. It's an easy enough change to add to all the drivers for this PR if that makes sense at all.

It's a fairly simple change, and would only be a BC break if someone was using "mail.endpoints.*" already in their configuration. At a guess I would say that's unlikely, but if an argument is made for it we could easily change the config key at this stage.

36864 committed:
Add ability to override default api endpoint

Fixes #23864 by allowing a user to set a "mail.endpoints.sparkpost" config key to choose between available endpoints found at https://developers.sparkpost.com/api/

tillkruss commented on src/Illuminate/Mail/Transport/SparkPostTransport.php:
```diff
@@ -54,7 +54,7 @@ public function send(Swift_Mime_SimpleMessage $message, &$failedRecipients = nul
         $message->setBcc([]);
 
-        $response = $this->client->post('https://api.sparkpost.com/api/v1/transmissions', [
+        $response = $this->client->post(config('mail.endpoints.sparkpost', 'https://api.sparkpost.com/api/v1/transmissions'), [
```
This should be `services.sparkpost.endpoint`.

GrahamCampbell replied:
You can't just call the foundation config function from the mail component. It doesn't exist.

nunomaduro replied:
What @GrahamCampbell means is: The helper `config()` is not available if you use the package `illuminate/mail` individually.

36864 committed:
Load API endpoint from options array

Previous approach assumed the existence of the `config()` helper.
This approach uses the the existing `options` configuration array in an attempt to maintain full BC while also making it clear that an endpoint can now be set.

36864 commented:
Thank you for the feedback. 

Changed the implementation to fetch the endpoint from the options array. This requires adding the options array with the endpoint element to the `services.sparkpost` configuration as such:

```
'sparkpost' => [
        'secret' => env('SPARKPOST_SECRET'),
	'options' => [
		'endpoint' => env('SPARKPOST_ENDPOINT', 'https://api.sparkpost.com/api/v1');
	]
],
```

This is the minimal change I could find to avoid using the config helper. No changes to method signatures or constructor parameters. Could also be extended to other Transports if necessary.

taylorotwell commented:
Your constructor doesn't make sense to me. What if endpoint isn't in the options?
